Build fail real

```
ERROR: test_subcase_errors (corehq.apps.app_manager.tests.test_build_errors.BuildErrorsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/test_build_errors.py", line 13, in test_subcase_errors
    errors = app.validate_app()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/models.py", line 3921, in validate_app
    raise ModuleIdMissingException
ModuleIdMissingException```

```
ERROR: test_exchange_snapshot_view (corehq.apps.domain.tests.test_views.TestDomainViews)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/tests/test_views.py", line 79, in test_exchange_snapshot_view
    follow=True
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/client.py", line 463, in post
    response = super(Client, self).post(path, data=data, content_type=content_type, **extra)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/client.py", line 297, in post
    return self.request(**r)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/client.py", line 424, in request
    six.reraise(*exc_info)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 115, in get_response
    response = callback(request, *callback_args, **callback_kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/utils/decorators.py", line 25, in _wrapper
    return bound_func(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/decorators.py", line 231, in _inner
    return view_func(request, domain_name, *args, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/utils/decorators.py", line 21, in bound_func
    return func(self, *args2, **kwargs2)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/views.py", line 287, in dispatch
    return super(BaseProjectSettingsView, self).dispatch(request, *args, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/utils/decorators.py", line 25, in _wrapper
    return bound_func(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/decorators.py", line 77, in _inner
    return view_func(req, domain_name, *args, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/utils/decorators.py", line 21, in bound_func
    return func(self, *args2, **kwargs2)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/views.py", line 169, in dispatch
    return super(LoginAndDomainMixin, self).dispatch(*args, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/views/generic/base.py", line 86, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/views.py", line 1691, in post
    copy_by_id=copy_by_id)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/models.py", line 806, in save_snapshot
    copy = self.save_copy(ignore=ignore, copy_by_id=copy_by_id)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/models.py", line 706, in save_copy
    comp = self.copy_component(res['value']['doc_type'], res['value']['_id'], new_domain_name, user=user)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/models.py", line 798, in copy_component
    new_doc.ensure_module_unique_ids(should_save=False)
AttributeError: 'UserRole' object has no attribute 'ensure_module_unique_ids'
```

--- a/corehq/apps/app_manager/exceptions.py
+++ b/corehq/apps/app_manager/exceptions.py
@@ -26,6 +26,10 @@ class AppEditingError(AppManagerException):
     pass
 
 
+class ModuleIdMissingException(AppManagerException):
+    pass
+
+
 class RearrangeError(AppEditingError):
     pass
 

--- a/corehq/apps/app_manager/models.py
+++ b/corehq/apps/app_manager/models.py
@@ -74,6 +74,7 @@ from .exceptions import (
     IncompatibleFormTypeException,
     LocationXpathValidationError,
     ModuleNotFoundException,
+    ModuleIdMissingException,
     RearrangeError,
     VersioningError,
     XFormException,
@@ -3931,7 +3932,8 @@ class Application(ApplicationBase, TranslationMixin, HQMediaMixin):
                 if xmlns_count[xmlns] > 1:
                     errors.append({'type': "duplicate xmlns", "xmlns": xmlns})
 
-        self.ensure_module_unique_ids(should_save=True)
+        if any(not module.unique_id for module in self.get_modules()):
+            raise ModuleIdMissingException
         modules_dict = {m.unique_id: m for m in self.get_modules()}
 
         def _parent_select_fn(module):

--- a/corehq/apps/app_manager/models.py
+++ b/corehq/apps/app_manager/models.py
@@ -3486,6 +3486,21 @@ class Application(ApplicationBase, TranslationMixin, HQMediaMixin):
             else:
                 map_item.version = self.version
 
+    def ensure_module_unique_ids(self, should_save=False):
+        """
+            Creates unique_ids for modules that don't have unique_id attributes
+            should_save: the doc will be saved only if should_save is set to True
+
+            WARNING: If called on the same doc in different requests without saving,
+            this function will set different uuid each time,
+            likely causing unexpected behavior
+        """
+        if any(not mod.unique_id for mod in self.modules):
+            for mod in self.modules:
+                mod.get_or_create_unique_id()
+            if should_save:
+                self.save()
+
     def create_app_strings(self, lang):
         gen = app_strings.CHOICES[self.translation_strategy]
         if lang == 'default':

--- a/corehq/apps/app_manager/views.py
+++ b/corehq/apps/app_manager/views.py
@@ -27,6 +27,7 @@ from corehq.apps.app_manager.exceptions import (
     FormNotFoundException,
     IncompatibleFormTypeException,
     ModuleNotFoundException,
+    ModuleIdMissingException,
     RearrangeError,
 )
 
@@ -812,6 +813,7 @@ def get_module_view_context_and_template(app, module):
 
         return options
 
+    # make sure all modules have unique ids
     app.ensure_module_unique_ids(should_save=True)
     if isinstance(module, CareplanModule):
         return "app_manager/module_view_careplan.html", {
@@ -2268,7 +2270,12 @@ def save_copy(request, domain, app_id):
     """
     comment = request.POST.get('comment')
     app = get_app(domain, app_id)
-    errors = app.validate_app()
+    try:
+        errors = app.validate_app()
+    except ModuleIdMissingException:
+        # For apps (mainly Exchange apps) that lost unique_id attributes on Module
+        app.ensure_module_unique_ids(should_save=True)
+        errors = app.validate_app()
 
     if not errors:
         try:

--- a/corehq/apps/app_manager/views.py
+++ b/corehq/apps/app_manager/views.py
@@ -785,14 +785,6 @@ def get_module_view_context_and_template(app, module):
     child_case_types = list(child_case_types)
     fixtures = [f.tag for f in FixtureDataType.by_domain(app.domain)]
 
-    def ensure_unique_ids():
-        # make sure all modules have unique ids
-        modules = app.modules
-        if any(not mod.unique_id for mod in modules):
-            for mod in modules:
-                mod.get_or_create_unique_id()
-            app.save()
-
     def get_parent_modules(case_type):
         parent_types = builder.get_parent_types(case_type)
         modules = app.modules
@@ -820,7 +812,7 @@ def get_module_view_context_and_template(app, module):
 
         return options
 
-    ensure_unique_ids()
+    app.ensure_module_unique_ids(should_save=True)
     if isinstance(module, CareplanModule):
         return "app_manager/module_view_careplan.html", {
             'parent_modules': get_parent_modules(CAREPLAN_GOAL),

--- a/corehq/apps/domain/models.py
+++ b/corehq/apps/domain/models.py
@@ -793,6 +793,8 @@ class Domain(Document, SnapshotMixin):
         if self.is_snapshot and doc_type == 'Application':
             new_doc.prepare_multimedia_for_exchange()
 
+        # when copying from app-docs that don't have
+        # unique_id attribute on Modules
         new_doc.ensure_module_unique_ids(should_save=False)
         new_doc.save()
         return new_doc

